src/main/java/frc/robot/subsystems/targetvision/TargetVision.kt

```diff
@@ -1,61 +1,203 @@
+                pose.add(measurement)
+                if ((measurement.timestamp - pose.first().timestamp > MAX_SAMPLE_TIME_MS)) { // TODO: find good value
+                    // cleanup old targets
+                    pose.removeFirst()
+                }
+                grouped = true
+                break
+            } else if((measurement.timestamp - pose.last().timestamp > MAX_SAMPLE_TIME_MS)){
+                robotPoses.remove(pose)
+            }
+        }
+
+        if(!grouped){
+            robotPoses.add(mutableListOf(measurement))
         }
+
+
+
+
+
+   }
+
+
+    //give a measurement a rating based on how likely it is to be our current primary target
+    private fun rateTrackingPreference(measurement: List<List<Measurement>>): Int{
+
+        fun getVelocity(samples: Pair<List<Measurement>, Int>): Double{
+            val tVector: List<Double> = samples.first.map{(timestamp, _) -> timestamp}
+            val txVector: List<Double> = samples.first.map{(_, target) -> target.tx}
+
+            // regressing tx
+            return tVector.zip(txVector).sumOf { pair -> pair.first * pair.second } /
+                    tVector.zip(tVector).sumOf { pair -> pair.first * pair.second }
+
+        }
+
+        var measurementIndexed = measurement.map { list -> Pair(list, measurement.indexOf(list))}
+
+        //filter by tx
+        measurementIndexed = measurementIndexed.sortedBy { pair -> abs(pair.first.last().pose.tx)}
+        val lowestTx = abs(measurementIndexed[0].first.last().pose.tx)
+        measurementIndexed = measurementIndexed.filter{ pair -> abs(pair.first.last().pose.tx) in lowestTx..lowestTx+TX_THRESHOLD}
+
+        //filter by distance
+        measurementIndexed = measurementIndexed.sortedBy {pair -> getDistance(pair.first.last().pose)}
+        val lowestDistance = getDistance(measurementIndexed[0].first.last().pose)
+        measurementIndexed = measurementIndexed.filter{ pair -> getDistance(pair.first.last().pose) in lowestDistance..lowestDistance + DISTANCE_THRESHOLD}
+
+
+        return measurementIndexed.minByOrNull { getVelocity(it) }!!.second
+
+
     }
 
-    private fun distanceToTarget(target: LimelightTarget_Retro): Double {
-        val angleToBucket = Units.degreesToRadians(CAMERA_PITCH.degrees + target.ty)
-        return (BUCKET_HEIGHT - LIMELIGHT_HEIGHT) / tan(angleToBucket)
+
+    private fun isBlueAlliance(): Boolean {
+        return DriverStation.getAlliance().equals(DriverStation.Alliance.Blue)
     }
 
-    // Constants
-    private val CAMERA_PITCH = Rotation2d.fromDegrees(110.0)
-    private val LIMELIGHT_HEIGHT = inchesToMeters(27.33)
-    private val BUCKET_HEIGHT = inchesToMeters(49.0) // TODO: this is a rough guess, measure exactly
-}
+    private val OpposingAlliance = if(isBlueAlliance()){Alliance.BLUE} else {Alliance.RED}
+
+     enum class Alliance(val id: Double){
+        RED(1.0),
+        BLUE(0.0)
+    }
+
+
+     private const val CAMERA_PITCH = 110 // degrees
```

Comment:
These should be in SI units. We don't want other units to be exposed to other parts of the code.